jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp

```diff
@@ -0,0 +1,272 @@
+#include "com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <string>
+#include <vector>
+#include <sys/time.h>
+#include <omp.h>
+
+#include "faiss/index_factory.h"
+#include "faiss/MetaIndexes.h"
+#include "faiss/index_io.h"
+#include "faiss/IndexHNSW.h"
+
+
+using std::string;
+using std::vector;
+
+std::unordered_map<string, faiss::MetricType> mapMetric = {
+        {"l2", faiss::METRIC_L2},
+        {"innerproduct", faiss::METRIC_INNER_PRODUCT}
```

Comment:
Doesn't FAISS also have cosine space?

Reply:
In Doc [Faiss-indexes](https://github.com/facebookresearch/faiss/wiki/Faiss-indexes) , i added inner product, which means If we have normalized vectors, the inner product becomes cosine similarity. 
And there is some issues talk about this: https://github.com/facebookresearch/faiss/issues/593

Reply:
Oh interesting. That makes sense.